### ISSUE TITLE
Add a --play option

### ISSUE THREAD

**Issue by phihag** [rg3/youtube-dl#628]
A common case is just wanting to watch the downloaded video. For that, we should add an option to play the video.

For starters, it can be a simple post-processing action. However, I think the user experience would be better if we'd have an option to either autoplay once a certain buffer (say, 2MB) has been written to disk. We may also tee the output to a pipe and pass _that_ to the video player in order for the video player not to stop playback once it reaches the current number of downloaded bytes.

**Comment by tbu-** [comment]
That feature would be really useful! Perhaps it would be useful to preload the next video while the previous one is still playing, so you don't get a pause.

**Comment by johnhawkinson** [comment]
@phihag, any updates on this? in 640 in 2013 you wrote:
> See #628 for an API that would allow playing videos while downloading them. I'm currently working on that and expect it to be finished in a week or so.

Obviously it's been more than a week :)

Since then `--exec` has been added, which is perhaps a model for how this should work from a UI perspective.

At the moment it's annoying to achieve with the tools we have available since you have to run youtube-dl twice: once to get the filename and once to actually do the download, and that overhead can be significant where getting the metadata is slow. (I suppose we could popen youtube-dl's output or use the api to get the filename while it runs the download...but at that point it seems much better to just add this option.)

Thanks for the update.

> We may also tee the output to a pipe and pass that to the video player in order for the video player not to stop playback once it reaches the current number of downloaded bytes.

I don't know that either of these are necessary. VLC, e.g., will continue playback on a regular file that is growing in realtime (at least when it grows faster than it plays), although it does get a little confused about the total time and present a pinned progress bar. But it's a better experience than playing from a pipe, since it cannot rewind or seek on a pipe.

**Comment by yan12125** [comment]
> since you have to run youtube-dl twice

Not sure why it's necessary for you. Here's an example with ```--exec```:
```
youtube-dl test:youtube --exec "mpv {}"
```

**Comment by johnhawkinson** [comment]
> Not sure why it's necessary for you. Here's an example with --exec:

I'm assuming we want to start playing immediately, not to wait for the download to finish. This is more important for much longer videos where the download might take a long time.

**Comment by yan12125** [comment]
Ah, I see your point.

> will continue playback on a regular file that is growing in realtime (at least when it grows faster than it plays)

Assume downloading is slower than playing. When mpv hits end of file, it quits normally.

**Comment by johnhawkinson** [comment]
If downloading is slower than playing, then this is a lot harder to think about, since the right time to start depends on speed and duration, and speed is not perfectly predictable, so there is no perfect solution. There's no universal correct answer for that case. 

But when the download is faster than playing, you want to start playing nearly instantaneously after the download starts. That's the case I'm interested in support for (and it's easier).

It starts to sound like the slower-than-realtime download calls for a pipe (maybe `--play 'sleep $DELAY; tail +1f {} | mpv'`), and the faster-than-realtime download calls for a filename (`--play 'vlc {}'`), but either way it does not change how this is implemented, it just changes what args the user calls  `--play` with.

Do you know what Philipp was intending? It doesn't seem very complicated...
The quick-and-dirty solution is to have FileDownloader.download() or FileDownloader.report_destination() fork, sleep for a second(?), and launch the player.  (I guess a snazzier subprocess solution is required for Windows...)

**Comment by yan12125** [comment]
The "sleep before start" does not fit all scenarios. Downloading speed may drop significantly in the middle, even if at first the speed looks quite fine.
I guess the best solution to this is asking players to add an option/feature, so that it waits in reaching the end of a file but not quits immediately.

**Comment by johnhawkinson** [comment]
I agree, solving slower-than-realtime download completely requires player support, and is beyond the scope of youtube-dl. But there is value in making this work where the download is faster than realtime, and we can do that today.

> I guess the best solution to this is asking players to add an option/feature, so that it waits in reaching the end of a file but not quits immediately.

Does mpv not do this? VLC does this when reading from a pipe.

**Comment by yan12125** [comment]
> But there is value in making this work where the download is faster than realtime, and we can do that today.

Yes. If players have proper supports, there's no need to handle slower-than-realtime cases specially in youtube-dl.

> Does mpv not do this? VLC does this when reading from a pipe.

mpv waits for pipes, too, while it quits for normal files. As you've mentioned, using pipe disables seeking, so I'd like to prevent it.

Other issues I can imagine as of now:

* The default format. As on-the-fly merging is not implemented yet, ```bestvideo+bestaudio``` is not applicable with ```--play```. Unfortunately, ```best``` has lower resolution on most YouTube videos and maybe more sites.
* When combining ```--play``` with other post processing options like ```--embed-thumbnail```, handling of temporary files is an issue, especially on Windows, in which it's disallowed to delete a file in use (e.g., opened by a player).

**Comment by johnhawkinson** [comment]
None of these seem like showstoppers.
(Also, `--play` should probably imply both `--no-part` and `--hls-use-mpegts`)

**Comment by yan12125** [comment]
Even if ```--no-part``` is specified, files are still re-generated with ffmpeg during thumbnail or subtitle embedding.

> None of these seem like showstoppers.

Can't agree. I bet there will be lots of complaints if ```--play``` does not give highest resolutions for YouTube videos.

**Comment by johnhawkinson** [comment]
So this should block until on-the-fly merging is implemented? (Is there an Issue #?)

It seems poor for "best" to be the enemy of "good." (Or, in this case "bestvideo+bestaudio" the enemy of `--play`.)

Have any of the players tried to solve the on-the-fly merging problem?

**Comment by yan12125** [comment]
> Is there an Issue #

Sorry I can't remember it.

> Have any of the players tried to solve the on-the-fly merging problem?

From https://github.com/mpv-player/mpv/commit/78caf6ae8634b9fe9589187d8febb927dce2ddeb, I guess someone added a feature in ffmpeg so that the ffmpeg-based mpv can handle bestvideo+bestaudio properly. I'm not sure what ffmpeg does is on-the-fly merging or not, though.
